manticore/native/heap_tracking/hook_malloc_library.py

```diff
@@ -0,0 +1,236 @@
+from manticore.native.state import State
+from manticore.native import Manticore
+from manticore.native.heap_tracking.malloc_lib_data import MallocLibData
+
+import logging
+
+logger = logging.getLogger(__name__)
+logger.setLevel(2)
+
+
+# Globals that will become class members to control amount of information being retrieved 
+# TODO(Sonya): fine tune these a little bit - need to be automated
+HOOK_SBRK_INFO = True
+HOOK_MMAP_INFO = True
+HOOK_MALLOC_RETURN = True
+HOOK_FREE_RETURN = True
+"""
+TODO(Sonya): class conversion 
+class hook_malloc_library:
+    def __init__(self, m: Manticore, sbrk:int, mmap:int, munmap:int, malloc:int, free:int, calloc=None, realloc=None, 
+        HOOK_SBRK_INFO: bool = True, HOOK_MMAP_INFO: bool = True, HOOK_MALLOC_RETURN: bool = True, HOOK_FREE_RETURN: bool = True):
+        self.sbrk = sbrk
+        self.mmap = mmap
+        self.munmap = munmap
+        self.malloc = malloc
+        self.free = free
+        
+"""
+
+
+def load_ret_addr(func: str, state: State):
+    """ Loads the return address of a function from the stack
+    (Assuming the next instruction to be executed is the start of a function call)
+    """
+    stack_location = state.cpu.read_register("RSP")
+    ret_addr = state.cpu.read_int(stack_location, state.cpu.address_bit_size)
+    logger.debug(f"Adding a hook for {func} callsite in state: {state.id}")
+    return ret_addr
+
+
+def hook_malloc_lib(initial_state: State, sbrk: int, mmap: int, munmap: int, malloc: int, free: int):
+    """ Function to add malloc hooks and do prep work
+    - TODO(Sonya): would like this to eventially be __init__() method for a class
+    once manticore hook callbacks have been debugged.
+    (from Eric) See: https://github.com/trailofbits/manticore/blob/master/tests/native/test_state.py#L163-L218
+    & https://github.com/trailofbits/manticore/blob/master/tests/native/test_state.py#L274-L278 to work on debugging this
+    """
+    initial_state.context["malloc_lib"] = MallocLibData()
+    
+    # Hook malloc and free
+    initial_state.add_hook(malloc, hook_malloc, after=False)
+    initial_state.add_hook(free, hook_free, after=False)
+    
+    initial_state.context['sbrk'] = sbrk
+    initial_state.context['mmap'] = mmap
+    initial_state.context['munmap'] = munmap
+
+
+def hook_mmap_return(state: State):
+    """ Hook to process munmap information and add a function hook to the callsite of munmap (which should 
+    be inside malloc or another function inside of malloc which calls munmap), post execution of the
+    munmap call.
+    """
+    ret_val = state.cpu.read_register("RAX")
```

Comment:
This register should be more generic, if possible

Reply:
I think there's a function that's used in the invoke_model code that I can use to replace some of the X86 specific reg names here